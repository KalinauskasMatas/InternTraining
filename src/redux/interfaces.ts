export interface Title {
  title: string;
}

export interface AuthFormInterface {
  isRegister: boolean;
}

export interface UserState {
  fname: string;
  surname: string;
  email: string;
  password: string;
}
