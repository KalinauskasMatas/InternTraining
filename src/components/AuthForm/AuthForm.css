.auth-form {
    background-color: var(--bright-background-color);
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    width: 250px;
    margin: auto;
    margin-bottom: 15px;
    border-radius: 30px;
}

.auth-form > label {
    margin: 15px 30px 3px 30px;
    font-size: 10px;
}

a {
    margin: auto;
}

form > input {
    margin: 0 30px 0 30px;
    border: none;
    background-color: var(--background-color);
    padding: 5px 10px;
    align-self: stretch;
}

.form-button {
    width: 125px;
    margin: 10px auto 12px auto;
    background-color: var(--dark-button-color);
    color: var(--bright-text-color);
    border: none;
    padding: 5px 35px;
    border-radius: 15px;
    cursor: pointer;
}

.error {
    height: 0.8rem;
    color: darkred;
    font-size: 0.8rem;
    margin: auto;
}