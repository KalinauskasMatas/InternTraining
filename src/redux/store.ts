import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/dist/query";
import { availableMoviesApi } from "./features/api/apiSlice";
// import availableMoviesSlice from "./features/availableMovies/availableMoviesSlice";
import currentUserReducer from "./features/currentUser/currentUserSlice";
import registeredUsersSlice from "./features/registeredUsers/registeredUsersSlice";

export const store = configureStore({
  reducer: {
    currentUser: currentUserReducer,
    registeredUsers: registeredUsersSlice,
    // availableMovies: availableMoviesSlice,
    [availableMoviesApi.reducerPath]: availableMoviesApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(availableMoviesApi.middleware),
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
