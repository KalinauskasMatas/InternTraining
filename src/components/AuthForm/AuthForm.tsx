import React, { useState } from "react";

import { useAppSelector, useAppDispatch } from "../../redux/hooks";

import { setCurrUser } from "../../redux/features/currentUser/currentUserSlice";
import { registerUser } from "../../redux/features/registeredUsers/registeredUsersSlice";

import { AuthFormInterface } from "../../interfaces";

import "./AuthForm.css";

const AuthForm = (props: AuthFormInterface) => {
  const [showRegister, setShowRegister] = useState(false);
  const [loginError, setLoginError] = useState("");
  const [registerError, setRegisterErrror] = useState("");

  const [loginData, setLoginData] = useState({
    email: "",
    password: "",
  });

  const [registerData, setRegisterData] = useState({
    fname: "",
    surname: "",
    email: "",
    emailRepeat: "",
    password: "",
    passwordRepeat: "",
  });

  const registeredUsers = useAppSelector((state) => state.registeredUsers);
  const dispatch = useAppDispatch();

  const revealRegister = () => {
    setShowRegister(!showRegister);
  };

  const handleLogin = (e: React.FormEvent) => {
    e.preventDefault();
    setLoginError("");

    const matchUser = registeredUsers.find(
      (user) =>
        user.email === loginData.email && user.password === loginData.password
    );

    if (!matchUser) {
      setLoginError("Invalid username or password");
      return;
    }

    dispatch(setCurrUser(matchUser));
  };

  const handleRegister = (e: React.FormEvent) => {
    e.preventDefault();
    setRegisterErrror("");

    const { fname, surname, email, emailRepeat, password, passwordRepeat } =
      registerData;

    if (email !== emailRepeat) {
      setRegisterErrror("The emails entered do not match");
      return;
    }

    if (password !== passwordRepeat) {
      setRegisterErrror("The passwords entered do not match");
      return;
    }

    if (registeredUsers.filter((user) => user.email === email).length > 0) {
      setRegisterErrror("User with this email already exists");
      return;
    }

    dispatch(
      registerUser({
        fname,
        surname,
        email,
        password,
        isAdmin: false,
        rentMovies: [],
      })
    );

    dispatch(
      setCurrUser({
        fname,
        surname,
        email,
        password,
        isAdmin: false,
        rentMovies: [],
      })
    );
  };

  const handleChange = (e: React.ChangeEvent<any>) => {
    if (e.target.parentNode.className.includes("login"))
      setLoginData({
        ...loginData,
        [e.target.id]: e.target.value,
      });
    else
      setRegisterData({
        ...registerData,
        [e.target.id]: e.target.value,
      });
  };

  if (!props.isRegister) {
    return (
      <article>
        <form className="login-form" onSubmit={handleLogin}>
          <label htmlFor="email">Email</label>
          <input
            type="email"
            id="email"
            placeholder="email"
            onChange={handleChange}
            required
          />
          <label htmlFor="password">Password</label>
          <input
            type="password"
            id="password"
            placeholder="password"
            onChange={handleChange}
            required
          />
          <input className="form-button" type="submit" value="Sign in" />
          <p className="error">{loginError}</p>
        </form>
      </article>
    );
  }
  return (
    <article>
      {showRegister ? (
        <form className="register-form" onSubmit={handleRegister}>
          <label htmlFor="fname">Name</label>
          <input
            type="text"
            id="fname"
            placeholder="name"
            minLength={2}
            onChange={handleChange}
            required
          ></input>
          <label htmlFor="surname">Surname</label>
          <input
            type="text"
            id="surname"
            placeholder="surname"
            minLength={2}
            onChange={handleChange}
          />
          <label htmlFor="email">Email</label>
          <input
            type="email"
            id="email"
            placeholder="email"
            onChange={handleChange}
            required
          />
          <label htmlFor="email-repeat">Email agail</label>
          <input
            type="email"
            id="emailRepeat"
            placeholder="email"
            onChange={handleChange}
            required
          />
          <label htmlFor="password">Password</label>
          <input
            type="password"
            id="password"
            placeholder="password"
            minLength={8}
            onChange={handleChange}
            required
          />
          <label htmlFor="passwordRepeat">Password again</label>
          <input
            type="password"
            id="passwordRepeat"
            placeholder="password"
            minLength={8}
            onChange={handleChange}
            required
          />
          <input className="form-button" type="submit" value="Register"></input>
          <p className="error">{registerError}</p>
        </form>
      ) : (
        <button className="form-button" onClick={revealRegister}>
          Register
        </button>
      )}
    </article>
  );
};

export default AuthForm;
