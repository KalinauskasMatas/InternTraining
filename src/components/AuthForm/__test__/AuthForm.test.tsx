import AuthForm from "../AuthForm";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";

import renderer from "react-test-renderer";

import { Provider } from "react-redux";
import { store } from "../../../redux/store";

afterEach(cleanup);

it("renders login without crashing", () => {
  render(
    <Provider store={store}>
      <AuthForm isRegister={false} />
    </Provider>
  );
});

it("renders register without crashing", () => {
  render(
    <Provider store={store}>
      <AuthForm isRegister={true} />
    </Provider>
  );
});

it("matches login snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <AuthForm isRegister={false} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("matches register snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <AuthForm isRegister={true} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
