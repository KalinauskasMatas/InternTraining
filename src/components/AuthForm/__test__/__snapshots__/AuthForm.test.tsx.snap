// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches login snapshot 1`] = `
<article>
  <form
    className="auth-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      id="email"
      onChange={[Function]}
      placeholder="email"
      required={true}
      type="email"
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      id="password"
      onChange={[Function]}
      placeholder="password"
      required={true}
      type="password"
    />
    <input
      className="form-button"
      type="submit"
      value="Sign In"
    />
    <p
      className="error"
    />
  </form>
</article>
`;

exports[`matches register snapshot 1`] = `
<article>
  <button
    className="form-button"
    onClick={[Function]}
  >
    Register
  </button>
</article>
`;
